src/main/java/org/mockito/internal/MockitoCore.java

```diff
@@ -60,11 +69,36 @@ public boolean isTypeMockable(Class<?> typeToMock) {
         }
         MockSettingsImpl impl = MockSettingsImpl.class.cast(settings);
         MockCreationSettings<T> creationSettings = impl.build(typeToMock);
+        checkDoNotMockAnnotation(creationSettings.getTypeToMock(), creationSettings);
         T mock = createMock(creationSettings);
         mockingProgress().mockingStarted(mock, creationSettings);
         return mock;
     }
 
+    private void checkDoNotMockAnnotation(Class<?> typeToMock, MockCreationSettings<?> creationSettings) {
+        checkDoNotMockAnnotationForType(typeToMock);
+        creationSettings.getExtraInterfaces().forEach(MockitoCore::checkDoNotMockAnnotationForType);
+    }
+
+    private static void checkDoNotMockAnnotationForType(Class<?> type) {
+        if (type == null) {
+            return;
+        }
+
+        if (SAFE_DONOTMOCK_ENFORCED_CLASSES.contains(type)) {
+            return;
+        }
+
+        DO_NOT_MOCK_ENFORCER.checkTypeForDoNotMockViolation(type).ifPresent(message -> {
+            throw new DoNotMockException(message);
+        });
+
+        checkDoNotMockAnnotationForType(type.getSuperclass());
+        Arrays.stream(type.getInterfaces()).forEach(MockitoCore::checkDoNotMockAnnotationForType);
+
+        SAFE_DONOTMOCK_ENFORCED_CLASSES.add(type);
```

Comment:
I feel uneasy about this perf tweak. Do we need it? Does it really make things faster? Static state that keeps growing with every mocked type may bite us later.

Did you find this perf tweak useful at G?

Reply:
I am pretty sure this is optimization is necessary as we can have large type hierarchies. I have asked internally for more background, but I expect this optimization to be necessary to not have a large performance impact for existing applications.

Reply:
Adding perf tweaks "on hunch", without evidence is not good for the codebase (growing complexity :-). I don't mind this change but I'd love to get your assurance that this is useful.

You have the approval so ship this change at will!

Reply:
I will double check before merging. I haven't had time during the Christmas vacation to do this work. I probably get to this late next week.